Add ignore threshold flag

arnaldo2792 opened the pull request:
*Issue #, if available:*
N / A

*Description of changes:*

```
240d1613 tuftool: add `ignore-threshold` flag
```

With [`18b7f740`](https://github.com/awslabs/tough/pull/188), a check was added to verify that the root.json file was signed with enough keys as specified in the thresholds. This is a problem when a cloud provider is used to store keys, and they are in different regions, since tuftool doesn't have the notion of regions while it fetches the keys.

With the `ignore-threshold` flag, `tuftool` will be more "permissive" and won't error out if a user attempts to sign the root.json file with fewer keys than specified in the thresholds.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

arnaldo2792 committed:
tuftool: add `ignore-threshold` flag

With `18b7f740`, a check was added to verify that the root.json file was
signed with enough keys as specified in the thresholds. This is a
problem when a cloud provider is used to store keys, and they are in
different regions, since tuftool doesn't have the notion of regions
while it fetches the keys.

With the `ignore-threshold` flag, `tuftool` will be more "permissive"
and won't error out if a user attempts to sign the root.json file with
fewer keys than specified in the thresholds.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

tjkirch commented on tuftool/src/root.rs:
```diff
@@ -106,6 +107,9 @@ pub(crate) enum Command {
         cross_sign: Option<PathBuf>,
+        ///Force signing with fewer keys than the specified threshold
+        #[structopt(short = "f")]
+        force: bool,
```
I think we should call this something more specific, because "force" could mean overriding any number of checks, and root roles are critical, so it's safer to override specific checks you know are OK to skip.  Something like `--ignore-threshold` or `--allow-keys-under-threshold` perhaps?

webern replied:
> I think we should call this something more specific, because "force" could mean overriding any number of checks, and root roles are critical, so it's safer to override specific checks you know are OK to skip. Something like `--ignore-threshold` or `--allow-keys-under-threshold` perhaps?

`--ignore-threshold` works for me (over `--allow-keys-under-threshold`).

tjkirch commented on tuftool/tests/root_command.rs:
```diff
@@ -84,7 +84,7 @@ fn add_key_all_roles(key: &str, root_json: &str) {
     Command::cargo_bin("tuftool")
         .unwrap()
-        .args(&["root", "sign", root_json, "-k", key])
+        .args(&["root", "sign", root_json, "-f", "-k", key])
```
Can we update testing so we have cases with fewer than the required number of keys, one without the new flag that we ensure fails, and one with the new flag that we ensure passes?

arnaldo2792 replied:
There is a test called `append_signature_root`, which initializes the root.json with root threshold 2, and uses `sign_root_json` with `-f`, which covers the the second part of the test that you ask. I'll the missing test case.

tjkirch commented on tuftool/src/root.rs:
```diff
@@ -327,12 +333,23 @@ impl Command {
+                        actual: keyids,
+                    });
+                }
+                // Print out a warning letting the user that the generated file isn't complient with
```
spelling: "compliant"

Can we clarify in the comments the difference between this check/warning and the one a few lines down?

tjkirch replied:
I see the top check was clarified to `Quick check that root is signed by enough key IDs, in all its roles`, but I'm still not sure what the lower check is for, that says `Signature check for root`?

tjkirch commented on .gitignore:
```diff
@@ -1,3 +1,3 @@
-/target
+**/target
```
I don't understand this change.  The tough repo has a Cargo.toml declaring the subdirectories to be in a workspace, and when you build (even from a subdirectory) the results go into `/target` in the root of the repo, not a subdirectory.  What are you doing when you see `target` somewhere else?

arnaldo2792 replied:
I ran `cargo run` inside `tuftool/`, and that generated `tuftool/target`:

```bash
Untracked files:
  (use "git add <file>..." to include in what will be committed)
        tough/target/
        tuftool/target/
```

tjkirch replied:
I don't have that behavior - `build` and `run` always put artifacts in the top-level `target`.  Let's talk offline about our setups.

arnaldo2792 replied:
I re-ran `cargo run` inside `tuftool`, I didn't see the folder. I'll remove the commit that updates the `.gitignore` file.

webern commented on tuftool/src/root.rs:
```diff
@@ -106,6 +107,9 @@ pub(crate) enum Command {
         ///Optional - Path of older root.json that contains the key-id
         #[structopt(short = "c", long = "cross-sign")]
         cross_sign: Option<PathBuf>,
+        ///Force signing with fewer keys than the specified threshold
+        #[structopt(short = "f")]
```
...just making sure: this is optional and defaults to `false`, right?

tjkirch replied:
That _seems_ to be the case from [structopt docs](https://docs.rs/structopt/0.3.23/structopt/#type-magic)... "`true` if the flag is present"...

webern reviewed (commented):
Looks good. Your CI checks failed due to a clippy lint.

`threshold: threshold` can be re-written as `threshold,` and clippy says you must...

zmrow commented on tuftool/src/root.rs:
```diff
@@ -352,10 +369,16 @@ impl Command {
+                    signature_count,
+                });
+            }
+            // Print out a warning letting the user that the generated file isn't complient with
```
This should probably say "letting the user *know* that the..."

Also - compliant is misspelled. :)

arnaldo2792 commented:
Forced push includes:

- Explanation about the first "quick check"
- Fixed typos
- Changed `force` to  `ignore-threshold`
- Add unit test to check for failures when `-i` isn't present.

arnaldo2792 commented:
Forced push includes:

- Fix linting failure

tjkirch commented on tuftool/tests/root_command.rs:
```diff
@@ -84,11 +84,19 @@ fn add_key_all_roles(key: &str, root_json: &str) {
     Command::cargo_bin("tuftool")
         .unwrap()
-        .args(&["root", "sign", root_json, "-k", key])
+        .args(&["root", "sign", root_json, "-i", "-k", key])
```
```suggestion
        // We don't have enough signatures to meet the threshold, so we have to pass `-i` to succeed
        .args(&["root", "sign", root_json, "-i", "-k", key])
```

tjkirch commented on tuftool/tests/root_command.rs:
```diff
@@ -84,11 +84,19 @@ fn add_key_all_roles(key: &str, root_json: &str) {
+fn sign_root_json_failure(key: &str, root_json: &str) {
+    Command::cargo_bin("tuftool")
+        .unwrap()
+        .args(&["root", "sign", root_json, "-k", key])
```
```suggestion
        // We don't have enough signatures to meet the threshold, so we should fail
        .args(&["root", "sign", root_json, "-k", key])
```

arnaldo2792 commented:
Forced push fixes nits in unit tests

zmrow reviewed (approved):
:volleyball:
